sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/streaming/StreamingReadSupport.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader.streaming;
+
+import org.apache.spark.sql.sources.v2.reader.ReadSupport;
+
+/**
+ * A base interface for streaming read support. This is package private and is invisible to data
+ * sources. Data sources should implement concrete streaming read support interfaces:
+ * {@link MicroBatchReadSupport} or {@link ContinuousReadSupport}.
+ */
+interface StreamingReadSupport extends ReadSupport {
+
+  /**
+   * Returns the initial offset for a streaming query to start reading from. Note that the
+   * streaming data source should not assume that it will start reading from its
+   * {@link #initialOffset()} value: if Spark is restarting an existing query, it will restart from
+   * the check-pointed offset rather than the initial one.
+   */
+  Offset initialOffset();
+
+  /**
+   * Deserialize a JSON string into an Offset of the implementation-defined offset type.
+   *
+   * @throws IllegalArgumentException if the JSON does not encode a valid offset for this reader
+   */
+  Offset deserializeOffset(String json);
```

Comment:
Why must this be JSON and why must it be a String? Why not byte[] and let the implementation choose the representation it prefers?

Reply:
Currently, there are two representations of any given offset: a connector-defined JVM object and a serialized JSON string.

Spark can't build the JVM object itself because it doesn't know what the right type is. If you know of some clean way for a connector to declare "here is the type of my offsets", we should do that instead, but I only know how to do it through reflection magic more confusing than the status quo.

I'd hesitate to introduce a third representation unless there's some concrete use case where JSON serialization won't work well.

Reply:
If Spark can serialize the offset based on the Offset interface, then it should deserialize to a generic Offset and use that to create a source-specific Offset. But then the Offset can't have any implementation-specific information, so why is the concrete class determined by the source anyway? This seems to me like a concrete case where the abstraction is failing.

Reply:
I think I understand what you're saying. I could get behind a proposal to simply define "arbitrary JSON string" as the one and only offset type, with each connector responsible for writing and parsing JSON however it'd like. All the existing offsets are trivial case classes anyway; it'd be a bit of a migration, but nothing architecturally difficult to handle.

I don't see how a `toBytes` method would help the problem. Neither arbitrary byte arrays nor arbitrary JSON strings let Spark know what type it's supposed to instantiate.

Reply:
I guess my core point is, we should stick with the existing serialization mechanism unless there's some kind of serialization we need to do which only a byte array can express. The serialization mechanism reaches deep into the execution layer, so coupling it with a connector API revamp is awkward.

Reply:
What I'm trying to say is this: either delegate serialization and deserialization entirely to the source -- in which case it passes you byte[] -- or standardize both the serialized representation and the class you will pass to the source.

I don't think it makes any sense for Spark to serialize to JSON and require that the source can deserialize it. Alternatively, if the source produces and consumes that JSON why are we forcing the source to use JSON?

Having a human-readable representation isn't a good justification for this because we have a ton of other objects that get serialized and aren't required to have a human-readable serialized form. Why is this different? If we care about the serialized representation so much, why allow customization of the Offset classes?

Reply:
As I said, I'm fine with defining arbitrary JSON strings as the single non-customizable offset type, if you think that would be better. (I think they would have to be strings, because making a JSON object the type would mean packaging some JSON library into the API.) I don't think it would ever be correct to have an Offset class which doesn't trivially reduce to a key-value map.

Reply:
@cloud-fan, I think this method should be removed from the API.

Reply:
This is an existing API, if we remove it, all the streaming sources can't work...

Reply:
Sorry, to be clear, this method was in the original streaming design doc https://docs.google.com/document/d/1VzxEuvpLfuHKL6vJO9qJ6ug0x9J_gLoLSH_vJL3-Cho which was sent out 2 months ago. If the reworking in this PR has made you realize there's a better way to do things, I think we should absolutely consider the alternative. But we can't just remove random methods without a complete proposal for what should replace them.

Reply:
@jose-torres, that's a fair point, this isn't a great place to decide whether serialization should be JSON or not.

It also looks like both serialization and deserialization is delegated to the source in that design doc, since Offset implementations must have a `json` method. That was a big part of my concern I pointed out, so requiring a source to provide both should be okay.